python/ray/tune/integration/torch.py

```diff
@@ -51,19 +53,52 @@ class _TorchTrainable(tune.Trainable):
     _function = None
     _num_workers = None
     _use_gpu = None
-    _num_cpus_per_worker = None
+    _gpus_per_worker = None
+    _cpus_per_worker = 1
+    _workers_per_node = None
+    _placement_group = None
 
-    __slots__ = ["workers", "_finished"]
+    __slots__ = ["workers", "_finished", "_placement_group"]
 
     @classmethod
     def default_process_group_parameters(self) -> Dict:
         return dict(timeout=timedelta(NCCL_TIMEOUT_S), backend="gloo")
 
-    @classmethod
+    @property
+    def worker_gpus(self) -> int:
+        if self._use_gpu:
+            return 0
+        return self._gpus_per_worker
+
+    @property
+    def worker_cpus(self) -> int:
+        return self._cpus_per_worker
+
+    @property
+    def should_colocate(self) -> bool:
+        return bool(self._workers_per_node)
+
+    @property
+    def num_hosts(self) -> Optional[int]:
+        if self.should_colocate:
+            return int(self._num_workers / self._workers_per_node)
+
     def get_remote_worker_options(self) -> Dict[str, int]:
-        num_gpus = 1 if self._use_gpu else 0
-        num_cpus = int(self._num_cpus_per_worker or 1)
-        return dict(num_cpus=num_cpus, num_gpus=num_gpus)
+        options = dict(num_cpus=self.worker_cpus, num_gpus=self.worker_gpus)
+        if self.should_colocate:
+            bundle = {
+                "CPU": self.worker_cpus * self._workers_per_node,
+                "GPU": self.worker_gpus * self._workers_per_node,
+            }
+            all_bundles = [bundle] * self.num_hosts
+            self._placement_group = placement_group(
```

Comment:
Do you really want STRICT_SPREAD? Why not SPREAD?

Reply:
Yeah, strict spread is a requirement. Imbalanced worker nodes can lead to timeouts or hangs when using NCCL2 to do collective communication.

Reply:
this is the opening of the rabbit hole https://github.com/NVIDIA/nccl/issues/19

Reply:
try a worker setup like [node=1, node=6, node=1], compared to [node=8] -> you'll see large performance hits.

Reply:
I see. So which is more preferable, [node=8] or [node=3, node=3, node=2]?

Reply:
resolved offline